### PR TITLE
refactor: generate application URL on backend side

### PR DESCRIPTION
**Pull request by mtojek**
Related: #8145

This PR delegates application URL generation from the site to the backend. Once it is merged, the generation will be performed in [one place](https://github.com/coder/coder/blob/main/coderd/httpapi/url.go#L33).

_Note: The last step of the issue will adjust the URL generation routine to prevent the bug of generating a too long subdomain._

### CODE REVIEW

File: coderd/workspaceagents.go [hunk 1]
```diff
@@ -64,8 +64,42 @@ func (api *API) workspaceAgent(rw http.ResponseWriter, r *http.Request) {
...
+	}
+	owner, err := api.Database.GetUserByID(ctx, workspace.OwnerID)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error fetching workspace owner.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
```
Comment (johnstcn) [inline coderd/workspaceagents.go:100]: Potential follow-up: would it make sense to modify `GetWorkspaceAppsByAgentID` to return the username and workspace name as well? Although this endpoint does not appear to be used by the UI at all, only by the CLI and unit tests.

File: coderd/workspaceagents.go [hunk 2]
```diff
@@ -1281,19 +1315,40 @@ func (api *API) workspaceAgentClientCoordinate(rw http.ResponseWriter, r *http.R
 	}
 }
 
-func convertApps(dbApps []database.WorkspaceApp) []codersdk.WorkspaceApp {
+// convertProvisionedApps converts applications that are in the middle of provisioning process.
+// It means that they may not have an agent or workspace assigned (dry-run job).
+func convertProvisionedApps(dbApps []database.WorkspaceApp) []codersdk.WorkspaceApp {
+	return convertApps(dbApps, database.WorkspaceAgent{}, database.User{}, database.Workspace{})
+}
+
```
Comment (johnstcn) [inline coderd/workspaceagents.go:1323]: nit: why have this extra function at all, you could simply just call `convertApps()` with the empty structs instead.

File: site/src/components/Resources/AppLink/AppLink.tsx [hunk 3]
```diff
@@ -51,9 +51,11 @@ export const AppLink: FC<AppLinkProps> = ({ app, workspace, agent }) => {
...
   }
 
-  if (appsHost && app.subdomain) {
-    const subdomain = `${appSlug}--${agent.name}--${workspace.name}--${username}`;
-    href = `${window.location.protocol}//${appsHost}/`.replace("*", subdomain);
+  if (appsHost && app.subdomain && app.subdomain_name) {
+    href = `${window.location.protocol}//${appsHost}/`.replace(
+      "*",
+      app.subdomain_name,
+    );
```
Comment (johnstcn) [inline site/src/components/Resources/AppLink/AppLink.tsx:58]: Might be good to add some unit tests for this component.

**Review by johnstcn — APPROVED** [review]
Note: this isn't technically a refactor, as we are changing the external-facing logic here.

Looks good to me, this particular endpoint does not appear to be called very often (I can only see it referenced by a few CLI commands), so the additional DB queries should not be an issue.

File: coderd/workspaceagents.go (hunk 2 above)
Comment (mtojek) [reply coderd/workspaceagents.go:1323]: That's true, I just personally find it "dirty" to call functions with multiple nil, nil, nil, {}, {}, {}, or "", "", "". 😅

File: coderd/workspaceagents.go (hunk 1 above)
Comment (mtojek) [reply coderd/workspaceagents.go:100]: I had that in mind as potential optimization, but as long it is not used, and could be deprecated, I will focus on deprecation instead 👍

File: site/src/components/Resources/AppLink/AppLink.tsx (hunk 3 above)
Comment (mtojek) [reply site/src/components/Resources/AppLink/AppLink.tsx:58]: I added a story test, but I can add an extra unit test as well 👍

Comment (mtojek) [reply site/src/components/Resources/AppLink/AppLink.tsx]: Done

**Review by johnstcn — APPROVED** [review]

**Pull request merged**